### ISSUE TITLE
Adaptive compression method for various column

### ISSUE THREAD

**Issue by sundy-li** [datafuselabs/databend#5238]
Let's create the task `Adaptive compression method for various column`

_Originally posted by @sundy-li in https://github.com/datafuselabs/databend/issues/5174#issuecomment-1120441514_

**Comment by DCjanus** [comment]
To start further discussion, here is my green though: 

add a byte to represent compression algorithm in the block, and try to compress block with lz4/zstd first, if result is larger than input, store data in plain.

Which means, each block would cost extra 1 byte at most.

**Comment by sundy-li** [comment]
Yes, I think the best compression algorithm can be learned from the data histogram distribution.

Let's split this task into two stages:

1. Users can manually indicate the compression method in SQL

2. Backgroud process task to train and learn the best compression method

**Comment by BohuTANG** [comment]
> Users can manually indicate the compression method in SQL

I think it is not a good idea for Databend, `no manual` is our design goal.

> Backgroud process task to train and learn the best compression method

Great.

The point to note is that Databend is a cloud data warehouse, the storage(like S3) is unlimited and cheap. The user's concern is(may) not the data size but the performance.

**Comment by dantengsky** [comment]
> 1. Users can manually indicate the compression method in SQL
  
  Maybe we should also take the encoding( or [specialized codec](https://clickhouse.com/docs/en/sql-reference/statements/create/table#specialized-codecs) for CH) into account as well so that we can define a column like this:
  
  `col DateTime Codec(RLE, LZ4)` ... 
  
  currently, the column encoding is PLAIN
 https://github.com/datafuselabs/databend/blob/eb9f43c9925c1f0c90e34fb91db02c253552c9f1/query/src/storages/fuse/io/write/block_writer.rs#L81-L92
   
   _comments in the code are outdated (parquet2 improved actively)_
